src/main.js

```diff
@@ -0,0 +1,379 @@
+      categories_copy_color.push(categoryHeaders[i]);
+    }
+    columns.push(categoryHeaders[i]);
+  }
+  // is there a time that exists in which categories_copy_color !== categories?
+  console.log("category_search_data:", category_search_data)
+  console.log("categories_copy_color:", categories_copy_color)
+  console.log("categories:", categories)
+  console.log("columns:", columns)
+  console.log("categoryHeaders:", categoryHeaders)
+
+  let dropdownBuilder = new DropdownBuilder();
+  dropdownBuilder.build(category_search_data, categories_copy_color, categories);
+  dropdownBuilder.setDropdownEventHandlers(redrawPlotWithoutZoom);
+  mainData = data;
+
+  // Initial plot draw happens here:
+  let needZoom = false;
+  highlighting(mainData, needZoom);
+};
+
+// getting header from csv file to make drowdown menus
+// NOTE: tsv() is an async function
+d3.tsv(dataset, loadMainData);
+
+
+function searchExactMatchEventHandler(event) {
+  if (document.getElementById("searchText").value) redrawPlotWithoutZoom();
+}
+
+function handleCheck1(event) {
+  if (document.getElementById("transpText").value) {
+    redrawPlotWithoutZoom();
+  }
+}
+
+function redrawPlotWithoutZoom() {
+  let needZoom = false;
+  highlighting(mainData, needZoom);
+}
+
+// it will be executed when (?? draw and) zoom button is pressed and the plot will zoomed out according to the points obtained by mouse click event
+function zoomEventHandler(){
+  if (plotOptionsReader.zoomCheckboxEnabled() === false) {
+    document.getElementById("zoomxy").value = ""; // clear the textbox
+  }
+  let needZoom = true;
+  highlighting(mainData, needZoom);
+}
+
+(function setEventHandlers() {
+  let zoomButton = plotOptionsReader.getZoomButton();
+  zoomButton.onclick = zoomEventHandler;
+
+  let colorOptions = plotOptionsReader.getColorOptions();
+  for (let i = 0; i < 2; i++) {
+    colorOptions[i].onclick = redrawPlotWithoutZoom;
+  };
+
+
+  let searchFormButton = plotOptionsReader.getSearchButton();
+  searchFormButton.addEventListener('click', (event) => {
+    event.preventDefault();
+    redrawPlotWithoutZoom();
+  });
+
+  let transparentSearchButton = plotOptionsReader.getTransparentSearchButton();
+  transparentSearchButton.addEventListener('click', (event) => {
+    event.preventDefault();
+    redrawPlotWithoutZoom();
+  });
+})();
+
+let coordinatesx = [];
+let coordinatesy = [];
+
+// function for plotting
+function highlighting(data, needZoom) {
+  let uniqueDataValuesToShape = [];
+  let spectrumGenerator;
+  var xValues = [], yValues = [], searchColumnValues = [];
+  var featureCategoryAndDataMap = {};
+  console.log('main data', data);
+
+  // remove the existing svg plot if any and clear side table
+  document.getElementById("demo3").innerHTML = "";
+  document.getElementById("predicted_words").innerHTML = "";
+  document.getElementById("frequent_words").innerHTML = "";
+  d3.select("svg").remove();
+  d3.select("table").remove();
+
+  // What is numerics doing?
```

Comment:
will remove this comment once I extract numerics to a separate module since the work associated with it only needs to be done once when the data is loaded.